packages/devtools_app/lib/src/service_manager.dart

```diff
@@ -564,24 +563,38 @@ class IsolateManager {
   }
 
   void _handleVmServiceClosed() {
+    cancel();
     _lastIsolateIndex = 0;
     _setSelectedIsolate(null);
     _isolateIndexMap.clear();
     _isolates.clear();
   }
+
+  void vmServiceOpened(VmServiceWrapper service) {
+    cancel();
+    _service = service;
+    autoDispose(service.onDebugEvent.listen(_handleDebugEvent));
+    autoDispose(service.onIsolateEvent.listen(_handleIsolateEvent));
+  }
 }
 
-class ServiceExtensionManager {
+class ServiceExtensionManager extends Disposer {
   VmServiceWrapper _service;
   IsolateManager _isolateManager;
 
-  bool _firstFrameEventReceived = false;
+  // Id so it is easy to avoid long running requests that were started
+  // for a previous vm service connection.
+  int vmServiceId = 0;
 
-  final Map<String, StreamController<bool>> _serviceExtensionController = {};
-  final Map<String, StreamController<ServiceExtensionState>>
-      _serviceExtensionStateController = {};
+  bool _checkForFirstFrameStarted = false;
+  bool get _firstFrameEventReceived => _firstFrameReceived.isCompleted;
+  Completer<void> _firstFrameReceived = Completer<void>();
+  Future<void> get firstFrameReceived => _firstFrameReceived.future;
 
-  final Map<String, ValueListenable<bool>> _serviceExtensionListenables = {};
+  final Map<String, ValueNotifier<bool>> _serviceExtensionAvailable = {};
```

Comment:
nit: not code you wrote, but put types on the right hand side, here and elsewhere

Reply:
done